src/ray/common/scheduling/scheduling_alg.cc

```diff
@@ -0,0 +1,204 @@
+      } else {
+        return -1;
+      }
+    } else {
+      if (task_req.custom_resources[i].req.demand > it->second.available) {
+        /// Resource constraint is violated.
+        if (task_req.custom_resources[i].req.soft) {
+          violations++;
+        } else {
+          return -1;
+        }
+      }
+    }
+  }
+  return violations;
+}
+
+int64_t ClusterResources::getSchedulableNode(const TaskReq &task_req, int64_t *violations) {
+  /// Min number of violations across all nodes that can schedule the request.
+  int64_t min_violations = INT_MAX;
+  /// Node associated to min_violations.
+  int64_t best_node = -1;
+  *violations = 0;
+
+  /// Check whether local node is schedulable.
+  auto it = nodes_.find(local_node_id_);
+  if (it != nodes_.end()) {
+    int64_t v = isSchedulable(task_req, it->second);
+    if (v == 0) {
+      auto it_p = task_req.placement_hints.find(local_node_id_);
+      if (it_p != task_req.placement_hints.end()) {
+        return local_node_id_;
+      }
+    }
+  }
+
+  /// Check whether any node in the request placement_hints, satisfes
+  /// all resource constraints of the request.
+  for (auto it_p = task_req.placement_hints.begin();
+       it_p != task_req.placement_hints.end(); ++it_p) {
+    auto it = nodes_.find(*it_p);
+    if (it != nodes_.end()) {
+      int64_t v = isSchedulable(task_req, it->second);
+      if (v == 0) {
+        return it->first;
+      }
+    }
+  }
+
+  int num_placement_hints = task_req.placement_hints.size();
+
+  for (auto it = nodes_.begin(); it != nodes_.end(); ++it) {
+    /// Return -1 if node not schedulable. otherwise return the number
+    /// of soft constraint violations.
+    int64_t v = isSchedulable(task_req, it->second);
+
+    /// If a hard constraint has been violated, ignnore this node.
+    if (v == -1) {
+      break;
+    }
+    // check wether placement_hints are satisfied
+    if (num_placement_hints > 0) {
+      auto it_p = task_req.placement_hints.find(it->first);
+      if (it_p == task_req.placement_hints.end()) {
+        /// Node not found in the placemen_hints list, so
+        /// record this a soft constraint violation.
+        v++;
+      }
+    }
+    // Update the node with the smallest number of soft constraints violated.
+    if (min_violations > v) {
+      min_violations = v;
+      best_node = it->first;
+    }
+    if (v == 0) {
+      *violations = 0;
+      return best_node;
+    }
+  }
+  *violations = min_violations;
+  return best_node;
+}
+
+bool ClusterResources::updateAvailableResources(int64_t node_id, const TaskReq &task_req) {
+  auto it = nodes_.find(node_id);
+  if (it == nodes_.end()) {
+    return false;
+  }
+  NodeResources &nr = it->second;
+
+  /// Just double check this node can still schedule the task request.
+  if (isSchedulable(task_req, nr) == -1) {
+    return false;
+  }
+
+  for (int i = 0; i < NUM_PREDIFINED_RESOURCES; i++) {
+    if (task_req.predefined_resources[i].demand > 0) {
```

Comment:
I think the entire `for` loop body can be replaced with:
`nr.capacities[i].available = std::max(0, nr.capacities[i].available - task_req.predefined_resources[i].demand);`